src/util/piped_process.h

```diff
@@ -0,0 +1,70 @@
+/// \file Subprocess communication with pipes
+///
+/// \author Diffblue Ltd.
+
+#ifndef CPROVER_UTIL_PIPED_PROCESS_H
+#define CPROVER_UTIL_PIPED_PROCESS_H
+
+class piped_processt
+{
+public:
+  /// Enumeration to keep track of child process state.
+  enum class process_statet
+  {
+    NOT_CREATED,
+    CREATED,
+    STOPPED,
+    ERROR
+  };
+
+  enum class process_send_responset
+  {
+    SUCCEEDED,
+    FAILED,
+    ERROR
+  };
+
+  /// Send a string message (command) to the child process.
+  /// \param message The string message to be sent.
+  process_send_responset send(const std::string &message);
+  /// Read a string from the child process' output.
+  std::string receive();
+  /// Wait until a string is available and read a string from the child
+  /// process' output.
+  std::string wait_receive();
+
+  /// Get child process status.
+  process_statet get_status();
+
+  /// See if this process can receive data from the other process.
+  /// \param timeout Amount of time to wait before timing out and
+  ///     returning: -1 is infiniute, 0 is non-blocking immediate.
```

Comment:
Using the magic value of `-1` is leaking the implementation detail, that `poll` uses `-1` to mean infinite waiting. I'd make this more expressive / API agnostic by using parameter name and type `optionalt<std::size_t> seconds_waiting_limit`. Note the use of an unsigned type to force users to use empty when they want unlimited waiting.

Reply:
Thanks for implementing this change.